File: src/builder/transpilers/latex/01-spacer.ts
import { hasClass } from '../_ast-utils.js';

/**
 * Transpiler de formato LaTeX: transforma Divs con clase .spacer (generados
 * por :: y :; en la capa semantica) a \vspace{\baselineskip}.
 *
 * Si el Div tiene ademas la clase .noindent (caso :;), agrega \noindent
 * al primer parrafo siguiente.
 *
 * Para sobrescribir en un proyecto, crear un archivo con el mismo
 * nombre en <proyecto>/transpilers/latex/01-spacer.ts.
 */

export const type = 'ast' as const;

export async function transform(ast: Record<string, unknown>): Promise<Record<string, unknown>> {
  const blocks = ast.blocks as unknown[];
  if (!Array.isArray(blocks)) return ast;

  const newBlocks: unknown[] = [];
  let pendingNoIndent = false;

  for (const block of blocks) {
    if (
      typeof block === 'object' &&
      block !== null &&
      (block as Record<string, unknown>).t === 'Div' &&
      hasClass(block as Record<string, unknown>, 'spacer')
    ) {
      newBlocks.push({ t: 'RawBlock', c: ['latex', '\\vspace{\\baselineskip}'] });
      pendingNoIndent = hasClass(block as Record<string, unknown>, 'noindent');
    } else if (pendingNoIndent && typeof block === 'object' && block !== null && (block as Record<string, unknown>).t === 'Para') {
      // Agregar \noindent al primer parrafo despues de un spacer noindent
      const para = block as Record<string, unknown>;
      const inlines = para.c as unknown[];
      if (Array.isArray(inlines)) {
        newBlocks.push({ ...para, c: [{ t: 'RawInline', c: ['latex', '\\noindent '] }, ...inlines] });
      } else {
        newBlocks.push(block);
      }
      pendingNoIndent = false;
    } else {
      newBlocks.push(block);
      pendingNoIndent = false;
    }
  }

  ast.blocks = newBlocks;
  return ast;
}

File: src/lib/resources/filters/latex/01-spacer.lua
-- Convierte Div.spacer en \vspace{\baselineskip} (formato LaTeX).
-- Si el Div tiene la clase noindent (caso :;), agrega \noindent al primer
-- párrafo siguiente (solo si es Para).
-- Uso: pandoc --from json --to latex --lua-filter latex/01-spacer.lua

local function has_class(block, cls)
  if block.t ~= 'Div' then return false end
  for _, c in ipairs(block.classes) do
    if c == cls then return true end
  end
  return false
end

-- Los spacers "::" se convierten en cualquier profundidad (items de lista,
-- celdas de tabla, blockquotes): sin este handler, el Div anidado se perdía
-- en silencio. Los spacers con noindent (:;) los gestiona Pandoc() a nivel
-- superior, porque el \noindent necesita el párrafo siguiente (hermano).
function Div(div)
  if not has_class(div, 'spacer') or has_class(div, 'noindent') then return nil end
  return pandoc.RawBlock('latex', '\\vspace{\\baselineskip}')
end

function Pandoc(doc)
  local blocks = {}
  local pending_noindent = false
  for _, block in ipairs(doc.blocks) do
    if has_class(block, 'spacer') then
      table.insert(blocks, pandoc.RawBlock('latex', '\\vspace{\\baselineskip}'))
      pending_noindent = has_class(block, 'noindent')
    elseif pending_noindent and block.t == 'Para' then
      table.insert(block.content, 1, pandoc.RawInline('latex', '\\noindent '))
      table.insert(blocks, block)
      pending_noindent = false
    else
      table.insert(blocks, block)
      pending_noindent = false
    end
  end
  doc.blocks = blocks
  return doc
end
